File: src/main/java/investments/controllers/InvestmentController.java
package investments.controllers;
import investments.BOLO.InvestmentForm;
import investments.db.DataAccess;
import investments.db.del.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.servlet.mvc.support.RedirectAttributes;

import javax.servlet.http.HttpServletRequest;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;

@Controller
@RequestMapping("/investments")
public class InvestmentController extends BaseController
{
    @Autowired
    private DataAccess dataAccess;

    public InvestmentController()
    {}    
    
    @RequestMapping(value="/{id}/view")
    public String showInvestment(Map<String, Object> model, @PathVariable Long id)
    {
        logger.info("Viewing investment: " + id);
        Investment investment = dataAccess.getInvestmentById(id);
        model.put("investment", investment);        
        return "/viewInvestment";
    }

    @RequestMapping(value="/{id}/delete", method = RequestMethod.GET)
    public String deleteInvestment(@PathVariable Long id)
    {
        logger.info("Deleting investment: " + id);
        dataAccess.deleteInvestmentbyId(id);
        return "redirect:/";
    }
   
    
    @RequestMapping(value="/populateDummyData", method = RequestMethod.GET)
    public String populateDummyData(Map<String, Object> model)
    {
        List<Investment> investments = dataAccess.getAllInvestments();
        List<AssetRegion> regions = dataAccess.getAllRegions();
        List<Risk> risks = dataAccess.getAllRisks();
        List<InvestmentGroup> groups = dataAccess.getAllGroups();
        List<InfluenceFactor> factors = dataAccess.getAllFactors();   
        
        Random r = new Random();
             
        // create dummy investments
        if(investments.isEmpty()){
            logger.info("no investments found, creating new data...");
            int investmentCount=10;
            for(int i = 0; i < investmentCount ;i++){
                Investment localInvestment = new Investment();
                localInvestment.setDesirabilityStatement("My desirability statement goes in here");
                localInvestment.setInitialInvestment(r.nextFloat());
                localInvestment.setName("My investment's name#"+i);
                localInvestment.setDescription("the description of investment goes in here");
                localInvestment.setSymbol("CTX");
                localInvestment.setValue(r.nextFloat());
                localInvestment.setValueProposition("The value proposition statement goes in here");
                
                                
                // Create dummy risks
                if(risks.isEmpty())
                {
                    risks.add(new Risk("Director dismissal", "Financial officer fired due to corruption", RiskType.Company));
                    risks.add(new Risk("Competition", "Competition from other companies", RiskType.Market));        
                    risks.add(new Risk("Fashion", "Fashion of the comodity", RiskType.Market));        
                    risks.add(new Risk("Earnings report", "Investor perception based on earnings", RiskType.Company));   
                    
                    risks.add(new Risk("", "", RiskType.Company));   
                }               
                
                // Create dummy groups
                if( groups.isEmpty()){
                    groups.add(new InvestmentGroup("Value Investments","High current p/e with potential to maintain."));
                    groups.add(new InvestmentGroup("Growth Investments","Low p/e with potential to grow"));
                    groups.add(new InvestmentGroup("Momentum Investments","Fashionalble trends"));
                    groups.add(new InvestmentGroup("Hybrid Investments","Bit of everything"));                        
                    
                    groups.add(new InvestmentGroup("Tactical","carefully considered group"));
                    groups.add(new InvestmentGroup("Strategic","Assets with a strategic goal associated with them"));
                    groups.add(new InvestmentGroup("Shares"," Equity in company shares - fractional part owner"));
                    groups.add(new InvestmentGroup("Gold","Commodity which is valuable"));
                    groups.add(new InvestmentGroup("Emerging markets"," places like Japan, Turkey, Brazil, Taiwan etc."));
                    
                }
                
                // Create dummy factors
                if(factors.isEmpty()){
                    InfluenceFactor factor1 = new InfluenceFactor("Weather","The climate will affect the investment.");
                    factor1.setInfluence("Sunny weather helps, rainy weather doesn't");
                    
                    InfluenceFactor factor2 = new InfluenceFactor("Competiion","The competition dictates te supply and demand");
                    factor2.setInfluence("The more cometition the less buiness you get if the competition or on par to you");
                       
                    String[] samples = new String[] { "Transport", 
                        "Travel/Tourism", 
                        "Utilities",
                        "Telecommunications",
                        "Professional Services/Consulting",
                        "Pharmaceutical/Medical Product",
                        "Oil/Gas",
                        "Mining/Metals",
                        "Manufacturing",
                        "IT (Hardware/Software/Services)",
                        "Investment Banking","Food and Beverage","Consumer Goods","Agriculture"};
                    
                    for( String each : samples)          
                    {
                        InfluenceFactor f = new InfluenceFactor(each,"description about "+each);
                        f.setInfluence("influenced by " + each);                        
                        factors.add(f);
                    }                   
                }
                
                Investment savedInvestment = dataAccess.saveInvestment(localInvestment);
                investments.add(savedInvestment);
                logger.info("Created investment name="+ localInvestment.getName());
                
                int gmax = 0;
                while(gmax == 0){
                    gmax = r.nextInt(groups.size());
                }
                for(int g = 0; g < gmax;g++){
                    InvestmentGroup group = groups.get(g);
                    group.addInvestment(savedInvestment);
                    savedInvestment.addGroup(group);
                }
                
                int fmax = 0;
                while(fmax == 0){
                    fmax = r.nextInt(factors.size());
                }                
                for(int f = 0; f < fmax;f++){
                    InfluenceFactor factor = factors.get(f);
                    factor.addInvestment(savedInvestment);                
                    savedInvestment.addFactor(factor);
                }
                
                int rmax = 0;
                while(rmax == 0){
                    rmax = r.nextInt(risks.size());
                }
                for( int rsk = 0; rsk < rmax;rsk++){
                    Risk risk = risks.get(rsk);
                    risk.addInvestment(savedInvestment);                
                    savedInvestment.addRisk(risk);         
                } 
                
                int regionMax = 0;
                while(regionMax == 0){ regionMax = r.nextInt(regions.size());}
                for( int reg = 0; reg < regionMax;reg++){
                    AssetRegion region = regions.get(reg);
                    region.addInvestment(localInvestment);
                    savedInvestment.addRegion(region);
                     dataAccess.saveRegion(region);
                }   
            }
        }

        //update investments with add groups, factors, risks etc.
        investments.forEach((investment) -> dataAccess.updateInvestment(investment));
        
        model.put("investments", investments);         
        return "investments";
    }
    
    @RequestMapping(value="/", method=RequestMethod.GET)
    public String home(Map<String, Object> model)
    {
        List<Investment> investments = dataAccess.getAllInvestments();
        model.put("investments", investments);         
        return "investments";
    }

    /***
     * This is take an investmentForm which is the result of the NewInvestment workflow and create the investment.
     * @param investmentForm the backing object the represents the new investment and its associated factors which will be created
     * @return the addInvestment page which commits it
     */
    @RequestMapping(method = RequestMethod.POST)
    public String submit(InvestmentForm investmentForm)
    {
        // Set up an investment object and populate it with details extracted from the new Investment(from the InvestmentForm)
        Investment investment = new Investment();

        // Extract the new investment from the investmentForm...
        investment.setName(investmentForm.getName());
        investment.setRegions(investmentForm.getRegions());
        investment.setInitialInvestment(investmentForm.getInitialInvestment());
        investment.setValueProposition(investmentForm.getValueProposition());
        investment.setDesirabilityStatement(investmentForm.getDesirabilityStatement());
        investment.setSymbol(investmentForm.getSymbol());
        
        dataAccess.saveInvestment(investment);
        
        logger.info("Creating investment: " + investmentForm.getName());
        return "redirect:/investments/";
    }   
    
    @RequestMapping(value = "/ShowAddInvestmentPage",method = RequestMethod.GET)
    public String addInvestmentPage(Map<String, Object> model)
    {        
        model.put("investment", new Investment());
        model.put("regions", dataAccess.getAllRegions());
        return "addinvestment";
    }

    @RequestMapping(value = "/update", method = RequestMethod.POST)
    public @ResponseBody
    String updateDisplayName(@RequestParam(value = "name") String name, @RequestParam(value = "value") String value, @RequestParam(value = "pk") String pk,  HttpServletRequest request) {
        Investment i = dataAccess.getInvestmentById(Long.parseLong(pk));
        switch(name){
            case "name":
                i.setName(value);
                break;
            case "description":
                i.setDescription(value);
                break;
            case "valueProposition":
                i.setValueProposition(value);
                break;

        }
        dataAccess.updateInvestment(i);
        return "";
    }
}

File: WinInvestmentTracker/WinInvestmentTracker/Controllers/InvestmentController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using WinInvestmentTracker.Common;
using WinInvestmentTracker.Models;
using WinInvestmentTracker.Models.DAL;

namespace WinInvestmentTracker.Controllers
{
    public class InvestmentController : Controller
    {
        // Access to our database
        ApplicationDbContext db = new ApplicationDbContext();
        
        public ActionResult Index()
        {
            var investments = db.Investments.ToList();
            return View(investments);
        }
        public ActionResult Api()
        {
            return Content("No API yet");
        }

        public ActionResult Create()
        {
            return View();
        }

        [HttpPost]
        public ActionResult Create(Investment investment)
        {
            try
            {
                //TODO: persit to database!
                return View("Details", investment);
            }
            catch
            {
                return View();
            }            
        }

        [RunThisAfterActionCompleteAttribute]        
        public ActionResult Details(Models.Investment investment)
        {            
            return View(investment);
        }

        public ActionResult PopulateDummyData()
        {
            var MAX = 10;
            var factors = new List<InvestmentInfluenceFactor> {
                    new InvestmentInfluenceFactor {
                        Name = "Weather", Description = "The climate will affect the investment.",
                        Influence = "Sunny weather helps, rainy weather doesn't"},
                    new InvestmentInfluenceFactor {
                        Name = "Competiion", Description = "The competition dictates te supply and demand",
                        Influence = "The more cometition the less buiness you get if the competition or on par to you"},                    
            };
            String[] samples = new String[]
            {
                    "Transport",
                    "Travel/Tourism",
                    "Utilities",
                    "Telecommunications",
                    "Professional Services/Consulting",
                    "Pharmaceutical/Medical Product",
                    "Oil/Gas",
                    "Mining/Metals",
                    "Manufacturing",
                    "IT (Hardware/Software/Services)",
                    "Investment Banking",
                    "Food and Beverage",
                    "Consumer Goods",
                    "Agriculture"
            };
            foreach (String each in samples)
            {
                InvestmentInfluenceFactor f = new InvestmentInfluenceFactor { Name = each, Description = "description about " + each };
                f.Influence = "influenced by " + each;
                factors.Add(f);
            }
            factors.ForEach(f => f.Investments = new List<Investment>());
            factors.ForEach(f => db.Factors.Add(f));
            db.SaveChanges();

            var groups = new List<InvestmentGroup> {
                new InvestmentGroup{ Name = "Value Investments", Description = "high current p/e with potential to maintain.", Type = "" },
                new InvestmentGroup{ Name = "Growth Investments", Description = "Low p/e with potential to grow", Type = "" },
                new InvestmentGroup{ Name = "Momentum Investments", Description = "Fashionalble trends", Type = "" },
                new InvestmentGroup{ Name = "Hybrid Investments", Description = "Bit of everything", Type = "" },
                new InvestmentGroup{ Name = "Tactical", Description = "carefully considered group", Type = "" },
                new InvestmentGroup{ Name = "Strategic", Description = "Assets with a strategic goal associated with them", Type = "" },
                new InvestmentGroup{ Name = "Shares", Description = "Equity in company shares - fractional part owner", Type = "" },
                new InvestmentGroup{ Name = "Gold", Description = "ommodity which is valuable when markets are volatile", Type = "" },
                new InvestmentGroup{ Name = "Emerging markets", Description = "places like Japan, Turkey, Brazil, Taiwan etc.", Type = "" },
            };

            groups.ForEach(g => g.Investments = new List<Investment>());
            groups.ForEach(g => db.Groups.Add(g));
            
            db.SaveChanges();

            var regions = new List<Region> {
                new Region { Name = "UK ALL COMPANIES (Fund Sector)" },
                new Region { Name = "UK SMALLER COMPANIES (Fund Sector)"},
                new Region { Name = "JAPAN (Fund Sector)"},
                new Region { Name = "JAPANESE SMALLER COMPANIES (Fund Sector)"},
                new Region { Name = "ASIA PACIFIC EXCLUDING JAPAN (Fund Sector)"},
                new Region { Name = "CHINA / GREATER CHINA SECTOR (Fund Sector)"},
                new Region { Name = "NORTH AMERICA (Fund Sector)"},
                new Region { Name = "NORTH AMERICAN SMALLER COMPANIES (Fund Sector)"},
                new Region { Name = "EUROPE INCLUDING UK (Fund Sector)"},
                new Region { Name = "EUROPE EXCLUDING UK (Fund Sector)"},
                new Region { Name = "EUROPEAN SMALLER COMPANIES (Fund Sector)"},
                new Region { Name = "GLOBAL (Fund Sector)"},
                new Region { Name = "GLOBAL EMERGING MARKETS (Fund Sector)" }
            };

            regions.ForEach(r => r.Investments = new List<Investment>());
            regions.ForEach(r => db.Regions.Add(r));
            db.SaveChanges();

            var risks = new List<InvestmentRisk> {
                new InvestmentRisk { Name = "Director dismissal", Description = "Financial officer fired due to corruption", Type = Common.RiskType.Company },
                new InvestmentRisk { Name = "Competition", Description = "Competition from other companies", Type = Common.RiskType.Company },
                new InvestmentRisk { Name = "Fashion", Description = "Fashion/popularity of the comodity", Type = Common.RiskType.Company },
                new InvestmentRisk { Name = "Earnings report", Description = "Investor perception based on earnings", Type = Common.RiskType.Company },
            };

            risks.ForEach(r => r.Investments = new List<Investment>());
            risks.ForEach(r => db.Risks.Add(r));
            db.SaveChanges();

            var investments = new List<Investment>();
            for (int i = 0; i < MAX; i++)
            {
                var investment = new Investment
                {
                    Description = "Description",
                    Symbol = "symbol",
                    DesirabilityStatement = "default desirabliity statement#" + i,
                    InitialInvestment = i,
                    Name = "investment#" + i,
                    Value = i,
                    ValueProposition = "value proposition#" + i
                };

                investment.Factors = new List<InvestmentInfluenceFactor>();
                investment.Groups = new List<InvestmentGroup>();
                investment.Regions = new List<Region>();
                investment.Risks = new List<InvestmentRisk>();                
                
                // now randonly assign some of the risks/factos/groups/regions to this investment

                int gmax = gmax = new Random(DateTime.Now.Second).Next(groups.Count);
                
                for (int g = 0; g < gmax; g++)
                {
                    InvestmentGroup group = groups[g];
                    group.Investments.Add(investment);
                    investment.Groups.Add(group);
                }

                int fmax = new Random(DateTime.Now.Second).Next(factors.Count);
                
                for (int f = 0; f < fmax; f++)
                {
                    InvestmentInfluenceFactor factor = factors[f];
                    factor.Investments.Add(investment);
                    investment.Factors.Add(factor);
                }

                int rmax = new Random(DateTime.Now.Second).Next(risks.Count);
                
                for (int rsk = 0; rsk < rmax; rsk++)
                {
                    InvestmentRisk risk = risks[rsk];
                    risk.Investments.Add(investment);
                    investment.Risks.Add(risk);
                }
                investments.Add(investment);
            }

            investments.ForEach(inv => db.Investments.Add(inv));
            db.SaveChanges();

            return RedirectToAction("Index");
        }
    }
}